runtime/asm/start_prototype.rs

```diff
@@ -22,41 +22,55 @@ struct RtHeader {
     bss_start: *mut u8,
 }
 
+mod class_id {
+    pub const COMMAND: usize = 2;
+    pub const MEMOP: usize = 5;
+    pub const EXIT: usize = 6;
+}
+
 #[link_section = ".start"]
 #[no_mangle]
-extern fn start_prototype(
+extern "C" fn start_prototype(
     rt_header: &RtHeader,
     _memory_start: usize,
     _memory_len: usize,
     _app_break: usize,
 ) -> ! {
     use crate::TockSyscalls;
-    use libtock_platform::{OneArgMemop, RawSyscalls, YieldType};
+    use libtock_platform::RawSyscalls;
 
     let pc: usize;
-    #[cfg(target_arch = "riscv32")]
     unsafe {
+        #[cfg(target_arch = "arm")]
+        asm!("mov {}, pc", lateout(reg) pc, options(nomem, nostack, preserves_flags));
+        #[cfg(target_arch = "riscv32")]
         asm!("auipc {}, 0", lateout(reg) pc, options(nomem, nostack, preserves_flags));
     }
     if pc != rt_header.start {
         // Binary is in an incorrect location: report an error via
-        // LowLevelDebug.
+        // LowLevelDebug then exit.
         unsafe {
-            TockSyscalls::four_arg_syscall(8, 1, 2, 0, 2);
-        }
-        // TODO: Replace with an Exit call when exit is implemented.
-        loop {
-            TockSyscalls::raw_yield(YieldType::Wait);
+            TockSyscalls::syscall4::<class_id::COMMAND>([
+                8 as *mut (),
```

Comment:
Part of me wonders if there wants to be some kind of `MiriInt` type to make it clear that these casts have real consequences.

Reply:
You can't use a `struct` as an `asm!` operand, which really limits what we can do here. We may be able to get away with a type alias, though. I do want to avoid having `*mut ()` appear everywhere, as it is very difficult to change that type later (it seems easy until you try to do it).

I may be able to improve this, but it would belong in a separate PR anyway. I'll keep thinking about it.

Reply:
Discussion moved to https://github.com/tock/libtock-rs/pull/289